File: sdk/typescript/src/patternDetection.ts
/**
 * Server-side pattern detection — temporal clusters, entity co‑occurrences,
 * topic clusters. Wraps the corresponding SpacetimeDB reducers.
 *
 * Each reducer reads from the memory table server-side and writes results to
 * a result table (compute‑and‑store pattern). The SDK wrapper calls the
 * reducer, then reads from the result table.
 */
import type { ClientLike } from "./types";

// ---------------------------------------------------------------------------
// Result types
// ---------------------------------------------------------------------------

export interface TemporalClusterRecord {
  id: string;
  workspace_id: string;
  start_time: number;
  end_time: number;
  count: number;
  /** JSON array of memory IDs */
  memory_ids: string;
  /** JSON array of top summary terms */
  summary_terms: string;
  created_at: number;
}

export interface EntityCooccurrenceRecord {
  id: string;
  workspace_id: string;
  entity_a: string;
  entity_b: string;
  count: number;
  strength: number;
  created_at: number;
}

export interface TopicClusterRecord {
  id: string;
  workspace_id: string;
  topic: string;
  count: number;
  /** JSON array of memory IDs */
  memory_ids: string;
  /** JSON array of top terms */
  top_terms: string;
  avg_confidence: number;
  created_at: number;
}

// ---------------------------------------------------------------------------
// Temporal Clusters
// ---------------------------------------------------------------------------

/**
 * Detect temporal clusters — groups of memories stored close together in time.
 * Uses 30‑minute buckets. Requires admin auth.
 * Returns list of cluster records sorted by start_time descending.
 */
export async function detectTemporalClusters(
  client: ClientLike,
  workspaceId: string,
): Promise<TemporalClusterRecord[]> {
  await client._call("detect_temporal_clusters", [workspaceId]);
  return client._sqlExec(
    "SELECT * FROM temporal_cluster_result WHERE workspace_id = :ws ORDER BY start_time DESC",
    { ws: workspaceId },
  ) as Promise<TemporalClusterRecord[]>;
}

// ---------------------------------------------------------------------------
// Entity Co‑occurrences
// ---------------------------------------------------------------------------

/**
 * Detect entity co‑occurrence patterns — pairs of entities that frequently
 * appear together in the same memory (from entities_json).
 * Requires admin auth. Returns list sorted by count descending.
 */
export async function detectEntityCooccurrences(
  client: ClientLike,
  workspaceId: string,
): Promise<EntityCooccurrenceRecord[]> {
  await client._call("detect_entity_cooccurrences", [workspaceId]);
  return client._sqlExec(
    "SELECT * FROM entity_cooccurrence_result WHERE workspace_id = :ws ORDER BY count DESC",
    { ws: workspaceId },
  ) as Promise<EntityCooccurrenceRecord[]>;
}

// ---------------------------------------------------------------------------
// Topic Clusters
// ---------------------------------------------------------------------------

/**
 * Detect topic clusters — groups of memories organised by shared term frequency.
 * Requires admin auth. Returns list sorted by cluster size (count) descending.
 */
export async function detectTopicClusters(
  client: ClientLike,
  workspaceId: string,
): Promise<TopicClusterRecord[]> {
  await client._call("detect_topic_clusters", [workspaceId]);
  return client._sqlExec(
    "SELECT * FROM topic_cluster_result WHERE workspace_id = :ws ORDER BY count DESC",
    { ws: workspaceId },
  ) as Promise<TopicClusterRecord[]>;
}

File: sdk/python/spacetime_memory/client/_pattern_detection.py
"""Server-side pattern detection mixin.

Wraps the SpacetimeDB server-side pattern detection reducers:
- detect_temporal_clusters
- detect_entity_cooccurrences
- detect_topic_clusters

Each reducer reads from the memory table server-side and writes results to
a result table (compute-and-store pattern). This mixin calls the reducer,
then reads from the result table.
"""

from __future__ import annotations

from typing import Any


class PatternDetectionMixin:
    """Mixin providing server-side pattern detection methods."""

    # ------------------------------------------------------------------
    # Temporal Clusters
    # ------------------------------------------------------------------

    def detect_temporal_clusters(
        self,
        workspace_id: str,
    ) -> list[dict[str, Any]]:
        """Detect temporal clusters — groups of memories stored close together in time.

        Uses 30-minute buckets server-side. Requires admin auth.

        Args:
            workspace_id: The workspace to analyse.

        Returns:
            List of cluster dicts with keys: id, workspace_id, start_time,
            end_time, count, memory_ids (JSON string), summary_terms (JSON string),
            created_at. Sorted by start_time descending.
        """
        self._call("detect_temporal_clusters", [workspace_id])
        return self._query("temporal_cluster_result", workspace_id=workspace_id)

    # ------------------------------------------------------------------
    # Entity Co‑occurrences
    # ------------------------------------------------------------------

    def detect_entity_cooccurrences(
        self,
        workspace_id: str,
    ) -> list[dict[str, Any]]:
        """Detect entity co‑occurrence patterns.

        Pairs of entities (from ``entities_json``) that frequently appear
        together in the same memory. Requires admin auth.

        Args:
            workspace_id: The workspace to analyse.

        Returns:
            List of co‑occurrence dicts with keys: id, workspace_id,
            entity_a, entity_b, count, strength, created_at.
            Sorted by count descending.
        """
        self._call("detect_entity_cooccurrences", [workspace_id])
        return self._query("entity_cooccurrence_result", workspace_id=workspace_id)

    # ------------------------------------------------------------------
    # Topic Clusters
    # ------------------------------------------------------------------

    def detect_topic_clusters(
        self,
        workspace_id: str,
    ) -> list[dict[str, Any]]:
        """Detect topic clusters — groups of memories organised by shared term frequency.

        Identifies top frequent terms across memories, then groups memories
        that share each top term. Requires admin auth.

        Args:
            workspace_id: The workspace to analyse.

        Returns:
            List of topic cluster dicts with keys: id, workspace_id,
            topic, count, memory_ids (JSON string), top_terms (JSON string),
            avg_confidence, created_at. Sorted by count descending.
        """
        self._call("detect_topic_clusters", [workspace_id])
        return self._query("topic_cluster_result", workspace_id=workspace_id)
